# Titanic

In [19]:
#Pandas is for dataframe manipulation
import pandas as pd 
#Scikit learn is used to build the model. It has various libraries like linear regression, decision tree etc.
from sklearn.linear_model import LogisticRegression
import numpy as np 
# import matplotlib.pyplot as plt #I did not use this
from sklearn.preprocessing import StandardScaler #used for standardization (sometimes called Z-score normalization)
# from sklearn.model_selection import train_test_split # Do not need this with the multiple datasets kaggle has given


In [20]:
#load the data into dataframes
#data retrieved from kaggle competitions
train = pd.read_csv("train.csv")
test_df = pd.read_csv("test.csv")
gender_submission = pd.read_csv("gender_submission.csv")

In [21]:
#will focus on train data for now. 
gender_submission.head(10)

PassengerId  Survived
0          892         0
1          893         1
2          894         0
3          895         0
4          896         1
5          897         0
6          898         1
7          899         0
8          900         1
9          901         0

In [22]:
#Merge the gender_submission database and the test_df database since they were given by kaggle
#but contain incomplete data in comparision to the train dataframe. This merge consolidates for further manipulation
test = pd.merge(test_df, gender_submission, on = "PassengerId", how="inner")
test.head()


PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  Survived  
0  34.5      0      0   330911   7.8292   NaN        Q         0  
1  47.0      1      0   363272   7.0000   NaN        S         1  
2  62.0      0      0   240276   9.6875   NaN        Q         0  
3  27.0      0      0   315154   8.6625   NaN        S         0  
4  22.0      1      1  3101298  12.2875   NaN        S         1

Going to need to apply any column changes to test and train. gender_submissions will be what we apply at the end with passenger IDs matching

In [23]:
# We need to turn the Pclass into its own columns in order to get a better prediction on who will be in what class
train['FirstClass'] = train.Pclass.apply(lambda p: 1 if p == 1 else 0)
test['FirstClass'] = test.Pclass.apply(lambda p: 1 if p == 1 else 0)
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  FirstClass  
0      0         A/5 21171   7.2500   NaN        S           0  
1      0          PC 17599  71.2833   C85        C           1  
2      0  STON/O2. 3101282   7.9250   NaN        S           0  
3      0            113803  53.1000  C123        S           1  
4      0            373450   8.0500   NaN        S           0

In [24]:
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  Survived  FirstClass  
0  34.5      0      0   330911   7.8292   NaN        Q         0           0  
1  47.0      1      0   363272   7.0000   NaN        S         1           0  
2  62.0      0      0   240276   9.6875   NaN        Q         0           0  
3  27.0      0      0   315154   8.6625   NaN        S         0           0  
4  22.0      1      1  3101298  12.2875   NaN        S         1           0

In [25]:
# Now the same thing for second class and then again for third class people. 
# We need to turn the Pclass into its own columns in order to get a better prediction on who will be in what class
train['SecondClass'] = train.Pclass.apply( lambda p: 1 if p == 2 else 0)
test['SecondClass'] = test.Pclass.apply( lambda p: 1 if p == 2 else 0)
train.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  FirstClass  SecondClass  
0      0         A/5 21171   7.2500   NaN        S           0            0  
1      0          PC 17599  71.2833   C85        C           1            0  
2      0  STON/O2. 3101282   7.9250   NaN        S           0            0  
3      0            113803  53.1000  C123        S           1            0  
4      0            373450   8.0500   NaN        S           0            0  
5      0            330877   8.4583   NaN        Q           0            0  
6      0             17463  51.8625   E46        S           1            0  
7      1            349909  21.0750   NaN        S           0            0  
8      2            347742  11.1333   NaN        S           0            0  
9      0            237736  30.0708   NaN        C           0            1

In [26]:
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  Survived  FirstClass  \
0  34.5      0      0   330911   7.8292   NaN        Q         0           0   
1  47.0      1      0   363272   7.0000   NaN        S         1           0   
2  62.0      0      0   240276   9.6875   NaN        Q         0           0   
3  27.0      0      0   315154   8.6625   NaN        S         0           0   
4  22.0      1      1  3101298  12.2875   NaN        S         1           0   

   SecondClass  
0            0  
1            0  
2            1  
3            0  
4            0

In [27]:
train['ThirdClass'] = train.Pclass.apply( lambda p: 1 if p == 3 else 0)
test['ThirdClass'] = test.Pclass.apply( lambda p: 1 if p == 3 else 0)
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  FirstClass  SecondClass  \
0      0         A/5 21171   7.2500   NaN        S           0            0   
1      0          PC 17599  71.2833   C85        C           1            0   
2      0  STON/O2. 3101282   7.9250   NaN        S           0            0   
3      0            113803  53.1000  C123        S           1            0   
4      0            373450   8.0500   NaN        S           0            0   

   ThirdClass  
0           1  
1           0  
2           1  
3           0  
4           1

In [28]:
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  Survived  FirstClass  \
0  34.5      0      0   330911   7.8292   NaN        Q         0           0   
1  47.0      1      0   363272   7.0000   NaN        S         1           0   
2  62.0      0      0   240276   9.6875   NaN        Q         0           0   
3  27.0      0      0   315154   8.6625   NaN        S         0           0   
4  22.0      1      1  3101298  12.2875   NaN        S         1           0   

   SecondClass  ThirdClass  
0            0           1  
1            0           1  
2            1           0  
3            0           1  
4            0           1

We know from stories of Titanic that women and children  can be saved in a life raft, but what about men? 
I think we should address gender and age next as it may help our model. 

In [29]:
# Let's make the sex cloumn into a binary column
train['Sex_binary'] = train.Sex.map({"male": 0, "female": 1}) 
test['Sex_binary'] = test.Sex.map({"male": 0, "female": 1})
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  FirstClass  SecondClass  \
0      0         A/5 21171   7.2500   NaN        S           0            0   
1      0          PC 17599  71.2833   C85        C           1            0   
2      0  STON/O2. 3101282   7.9250   NaN        S           0            0   
3      0            113803  53.1000  C123        S           1            0   
4      0            373450   8.0500   NaN        S           0            0   

   ThirdClass  Sex_binary  
0           1           0  
1           0           1  
2           1           1  
3           0           1  
4           1           0

I could of done this earlier, but in this data, I am not likely to use the following columns right now. I may need them later for further analysis but at present they are just cluttering up my dataset. Lets just drop them for now from each dataframe.

Pclass, Name, SibSp, Parch, Ticket, Fare, Cabin, Embarked

In [30]:
columns_to_drop = ["Pclass", "Name", "Sex", "SibSp", "Parch", "Ticket", "Fare", "Cabin", "Embarked"]

train = train.drop(columns_to_drop, axis = 1)
test = test.drop(columns_to_drop, axis = 1)
train.head()

PassengerId  Survived   Age  FirstClass  SecondClass  ThirdClass  \
0            1         0  22.0           0            0           1   
1            2         1  38.0           1            0           0   
2            3         1  26.0           0            0           1   
3            4         1  35.0           1            0           0   
4            5         0  35.0           0            0           1   

   Sex_binary  
0           0  
1           1  
2           1  
3           1  
4           0

In [31]:
train.describe()

PassengerId    Survived         Age  FirstClass  SecondClass  \
count   891.000000  891.000000  714.000000  891.000000   891.000000   
mean    446.000000    0.383838   29.699118    0.242424     0.206510   
std     257.353842    0.486592   14.526497    0.428790     0.405028   
min       1.000000    0.000000    0.420000    0.000000     0.000000   
25%     223.500000    0.000000   20.125000    0.000000     0.000000   
50%     446.000000    0.000000   28.000000    0.000000     0.000000   
75%     668.500000    1.000000   38.000000    0.000000     0.000000   
max     891.000000    1.000000   80.000000    1.000000     1.000000   

       ThirdClass  Sex_binary  
count  891.000000  891.000000  
mean     0.551066    0.352413  
std      0.497665    0.477990  
min      0.000000    0.000000  
25%      0.000000    0.000000  
50%      1.000000    0.000000  
75%      1.000000    1.000000  
max      1.000000    1.000000

In [32]:
train["Age"].count()
# we have 714 rows with an age. So we have some missing data. We have 891 rows 

714

## Types of missing data:
- **Missing Completely at Random (MCAR):** No pattern to the missingness.
- **Missing at Random (MAR):** Missingness might be related to other observed variables, but not the missing value itself.
- **Missing Not at Random (MNAR):** The missingness depends on the value of the missing data itself.

## Strategies for Handling Missing Values

- **Listwise Deletion:** Remove rows containing missing values. Use with caution, especially if you have a large number of missing entries, as you could lose valuable information.
- **Pairwise Deletion:** Utilize all available data, but computations might involve different subsets of data.
- **Imputation:** Fill in missing values with estimated substitutes. **Common methods:**
- **Mean/Median Imputation:** Replace missing values with the mean/median of the column. Suitable for numerical data.
- **Mode Imputation:** Replace missing categorical values with the most frequent category.
- **Predictive Modeling:** Create a model to predict missing values based on other variables. This can be more sophisticated.


This is an exert summary from google's Gemini

In [33]:
#Let's fill in the ages with the mean of all ages.
train['Age'].fillna(value = round(train['Age'].mean()), inplace = True) #look up .fillna function
test['Age'].fillna(value = round(test['Age'].mean()), inplace = True) 
train["Age"].count() #now we have every row accounted for. 

891

In [35]:
test.describe()


PassengerId         Age    Survived  FirstClass  SecondClass  \
count   418.000000  418.000000  418.000000  418.000000   418.000000   
mean   1100.500000   30.216507    0.363636    0.255981     0.222488   
std     120.810458   12.635016    0.481622    0.436934     0.416416   
min     892.000000    0.170000    0.000000    0.000000     0.000000   
25%     996.250000   23.000000    0.000000    0.000000     0.000000   
50%    1100.500000   30.000000    0.000000    0.000000     0.000000   
75%    1204.750000   35.750000    1.000000    1.000000     0.000000   
max    1309.000000   76.000000    1.000000    1.000000     1.000000   

       ThirdClass  Sex_binary  
count  418.000000  418.000000  
mean     0.521531    0.363636  
std      0.500135    0.481622  
min      0.000000    0.000000  
25%      0.000000    0.000000  
50%      1.000000    0.000000  
75%      1.000000    1.000000  
max      1.000000    1.000000

# Supervised Learning

- We will use a features and a label to train these types of models. 


In [60]:
#I want to focus on training a model on Age, Sex_binary, FirstClass, SecondClass, ThirdClass 
#The goal is to predict whether or not the user survived based on this. 
train_features = train[["Age", "Sex_binary", "FirstClass", "SecondClass", "ThirdClass"]]
train_labels = train["Survived"]
test_features = test[["Age", "Sex_binary", "FirstClass", "SecondClass", "ThirdClass"]]
test_labels = gender_submission["Survived"]
# gender_submission.head()

In [31]:
# Scale the feature data so it has mean = 0 and standard deviation = 1
scaler = StandardScaler()
train_features = scaler.fit_transform(train_features)
test_features = scaler.transform(test_features)

In [32]:
model = LogisticRegression()
model.fit(train_features, train_labels)

C:\Users\pc\AppData\Roaming\Python\Python310\site-packages\sklearn\utils\validation.py:1111: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression()

Not sure what the error is. Will need to go back and look more into it but it still runs for now.

In [33]:
print(model.score(train_features, train_labels))

0.8013468013468014


In [35]:
#lets see the coefficents -  Age, Sex_binary, FirstClass, SecondClass, ThirdClass 
print(model.coef_)


[[-0.42603339  1.23609343  0.54127538  0.062455   -0.51719478]]


Coefficients represent the relationship between each feature (variable) and the log-odds of the target event

A positive coefficient means that an increase in the feature's value is associated with a higher likelihood of survival.

A negative coefficient means that an increase in the feature's value is associated with a lower likelihood of survival.

- Age = -0.42603339 
- Sex_binary = 1.23609343      
- FirstClass = 0.54127538 
- SecondClass = 0.062455   
- ThirdClass = -0.51719478


In [39]:
#I remember when I did this 2 years ago, we used Jack and Rose then ourselves to make predictions on the model and we mad ethem in a np.array
Jack = np.array([20.0, 0.0, 0.0, 0.0, 1.0])
Rose = np.array([17.0, 1.0, 1.0, 0.0, 0.0])
Dom = np.array([29.0,  0.0, 0.0, 1.0, 0.0])

In [46]:
passenger_predict = np.array([Jack, Rose, Dom])

In [47]:
passenger_predict = scaler.transform(passenger_predict)

C:\Users\pc\AppData\Roaming\Python\Python310\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


That warning again?..

In [49]:
#prediction time! My favorite part
# Make survival predictions!
print(model.predict(passenger_predict)) #This will print a 1 or 0 for surivied or did not survive 
print(model.predict_proba(passenger_predict)) #this will give us how likely for each option

[0 1 0]
[[0.88542238 0.11457762]
 [0.05002647 0.94997353]
 [0.75873431 0.24126569]]


Jack had an 88.5% of NOT surviving based on the data. 
Rose had a 95% chance of surviving. 
Dom would of had a 75.9% chance of NOT surviving.